### PR TITLE
Enable allocatable support for Windows nodes

### LINKED ISSUE #69938: enforce-node-allocatable=kube-reserved/system-reserved don't work on Windows
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

After running some tests, we're seeing that the combined usage of kubelet + docker can be around 2Gi. We need to be able to account for this in scheduling to avoid crashing these services. The first step towards this is using kube-reserved and system-reserved to avoid oversubscribing the virtual memory space available on the Windows nodes with too many containers.

**What you expected to happen**:

It should reduce node allocatable memory by kube-reserved + system-reserved.

**How to reproduce it (as minimally and precisely as possible)**:

`kubelet.exe … --enforce-node-allocatable=kube-reserved --kube-reserved=memory=2Gi`
`kubelet.exe … --enforce-node-allocatable=system-reserved --system-reserved=memory=2Gi`

This logs an error in the kubelet log such as:

```
Flag --kube-reserved has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
F1017 17:21:13.974464   77496 server.go:207] [invalid configuration: EnforceNodeAllocatable (--enforce-node-allocatable) is not supported unless CgroupsPerQOS (--cgroups-per-qos) feature is turned on, invalid configuration: EnforceNodeAllocatable (--enforce-node-allocatable) [kube-reserved] is not supported on Windows]
```

code: 
https://github.com/kubernetes/kubernetes/blob/688550b2d34b20474f98c0c967bd846042141182/pkg/kubelet/apis/config/validation/validation_windows.go#L37

**Anything else we need to know?**:
/milestone v1.13
/assign feiskyer
/sig windows

This blocks a PR where I want to set it by default to 2Gi https://github.com/Azure/acs-engine/pull/3999

**Environment**:
- Kubernetes version (use `kubectl version`): v1.11.3 (also v1.12)
- Cloud provider or hardware configuration: Azure
- OS (e.g. from /etc/os-release): Windows Server version 1803
- Install tools: acs-engine

### PR DESCRIPTION
**Pull request by feiskyer**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR enables kubelet options `--system-reserved` and `--kube-reserved` on Windows nodes.

Without this change, Windows nodes' allocatable are always same as capacity, even if you have set both options, e.g. `--system-reserved=cpu=100m,memory=100Mi,ephemeral-storage=1Gi --kube-reserved=cpu=1200m,memory=100Mi,ephemeral-storage=1Gi`:

```
Capacity:
 attachable-volumes-azure-disk:  4
 cpu:                            2
 ephemeral-storage:              0
 memory:                         8388148Ki
 pods:                           110
Allocatable:
 attachable-volumes-azure-disk:  4
 cpu:                            2
 ephemeral-storage:              0
 memory:                         8388148Ki
 pods:                           110
```

After this change, allocatable would be `capacity- system-reserverd - kube-reserved`:

```
Capacity:
 attachable-volumes-azure-disk:  4
 cpu:                            2
 ephemeral-storage:              268075004Ki
 memory:                         8388148Ki
 pods:                           110
Allocatable:
 attachable-volumes-azure-disk:  4
 cpu:                            700m
 ephemeral-storage:              244910439630
 memory:                         8080948Ki
 pods:                           110
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #69938

**Special notes for your reviewer**:

Note that `--enforce-node-allocatable` is still not supported because there's no cgroups support on Windows.

**Does this PR introduce a user-facing change?**:

```release-note
kubelet --system-reserved and --kube-reserved are supported now on Windows nodes
```

### CODE REVIEW

**Comment by feiskyer** [comment]
Kindly ping @yujuhong @PatrickLang. Could you help to take a look at the PR?

**Comment by PatrickLang** [comment]
/lgtm
Still need to do some testing on it though. @yujuhong or @pjh do you want to take a look?

**Comment by pjh** [comment]
Sounds like a nice improvement! I have nothing to add to the review. I'm not quite in a position to help test this yet, our automated tests on GCE are coming along but not fully ready.

**Comment by feiskyer** [comment]
@pjh  Thanks.

ping @yujuhong. Could you have a look at this?

**Comment by yujuhong** [comment]
Looks good overall with some nits.
Will approve once they are addressed (since the PR has already been lgtm'd).

File: pkg/kubelet/cm/node_container_manager.go [hunk 1]
```diff
@@ -213,30 +211,6 @@ func (cm *containerManagerImpl) GetNodeAllocatableReservation() v1.ResourceList
 	return result
 }
 
-// hardEvictionReservation returns a resourcelist that includes reservation of resources based on hard eviction thresholds.
```
[the remark this answers is not in the document]
Comment (feiskyer) [reply pkg/kubelet/cm/node_container_manager.go]: ACK

File: pkg/kubelet/cm/container_manager_windows.go [hunk 2]
```diff
@@ -19,28 +19,149 @@ limitations under the License.
...
+func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*kubecontainer.RunContainerOptions, error) {
+	return &kubecontainer.RunContainerOptions{}, nil
+}
+
+func (cm *containerManagerImpl) UpdatePluginResources(*schedulercache.NodeInfo, *lifecycle.PodAdmitAttributes) error {
+	return nil
+}
+
+func (cm *containerManagerImpl) InternalContainerLifecycle() InternalContainerLifecycle {
+	return &internalContainerLifecycleImpl{cpumanager.NewFakeManager()}
```
[the remark this answers is not in the document]
Comment (feiskyer) [reply pkg/kubelet/cm/container_manager_windows.go]: This is actually same with stub container manager, which is only for conforming to the interfaces, but without actual actions.

File: pkg/kubelet/cm/container_manager_windows.go [hunk 3]
```diff
@@ -19,28 +19,149 @@ limitations under the License.
 package cm
 
 import (
-	"github.com/golang/glog"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (feiskyer) [reply pkg/kubelet/cm/container_manager_windows.go]: ACK

**Comment by feiskyer** [comment]
@yujuhong Addressed comments. PTAL

**Pull request merged**